// @cscfi/csc-ui/dist/collection/components/c-link/c-link.js
import { h } from "@stencil/core";
export class CLink {
  constructor() {
    this.href = null;
    this.underline = false;
    this.target = null;
    this.weight = '600';
  }
  render() {
    const classList = {
      underline: this.underline,
    };
    const style = {
      fontWeight: this.weight.toString(),
    };
    return (h("a", { class: classList, href: this.href, target: this.target, style: style }, h("slot", null)));
  }
  static get is() { return "c-link"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["c-link.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-link.css"]
    };
  }
  static get properties() {
    return {
      "href": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Url of link"
        },
        "attribute": "href",
        "reflect": false,
        "defaultValue": "null"
      },
      "underline": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Display line under the link"
        },
        "attribute": "underline",
        "reflect": false,
        "defaultValue": "false"
      },
      "target": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "regular target attribute of a hyperlink"
        },
        "attribute": "target",
        "reflect": false,
        "defaultValue": "null"
      },
      "weight": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Customisable font weight"
        },
        "attribute": "weight",
        "reflect": false,
        "defaultValue": "'600'"
      }
    };
  }
}
