// @cscfi/csc-ui/dist/collection/components/c-login-button/c-login-button.js
import { h } from "@stencil/core";
export class CLoginButton {
  constructor() {
    this.href = '';
    this.src = undefined;
    this.alt = '';
  }
  render() {
    const props = Object.assign({ tabindex: '0' }, (!!this.href ? { href: this.href } : {}));
    return (h("a", Object.assign({}, props), h("img", { src: this.src || '', alt: this.alt }), h("div", null, h("slot", null))));
  }
  static get is() { return "c-login-button"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["c-login-button.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-login-button.css"]
    };
  }
  static get properties() {
    return {
      "href": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [],
          "text": "Login provider link. Do not set if using a javascript click handler"
        },
        "attribute": "href",
        "reflect": false,
        "defaultValue": "''"
      },
      "src": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": true,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Login provider logo url"
        },
        "attribute": "src",
        "reflect": false
      },
      "alt": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Alt description for logo"
        },
        "attribute": "alt",
        "reflect": false,
        "defaultValue": "''"
      }
    };
  }
}
