// @cscfi/csc-ui/dist/collection/components/c-radio/c-radio.js
import { Host, h } from "@stencil/core";
export class CRadio {
  constructor() {
    this.checked = false;
    this.disabled = false;
    this.value = undefined;
  }
  render() {
    return h(Host, null);
  }
  static get is() { return "c-radio"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["c-radio.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-radio.css"]
    };
  }
  static get properties() {
    return {
      "checked": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set option as checked"
        },
        "attribute": "checked",
        "reflect": false,
        "defaultValue": "false"
      },
      "disabled": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Disable the radio button"
        },
        "attribute": "disabled",
        "reflect": false,
        "defaultValue": "false"
      },
      "value": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Radio button value"
        },
        "attribute": "value",
        "reflect": false
      }
    };
  }
}
