// @cscfi/csc-ui/dist/collection/components/c-row/c-row.js
import { h, Host } from "@stencil/core";
export class CRow {
  constructor() {
    this.gap = 0;
    this.nowrap = false;
    this.align = undefined;
    this.justify = undefined;
  }
  componentDidLoad() {
    this.el.style.setProperty('--_c-row-gap', this.el.dataset.gap);
  }
  render() {
    const classes = {
      'c-row': true,
      'c-row--nowrap': this.nowrap,
      [`c-row--align-${this.align}`]: !!this.align,
      [`c-row--justify-${this.justify}`]: !!this.justify,
    };
    return (h(Host, { "data-gap": `${this.gap}px`, class: classes }, h("slot", null)));
  }
  static get is() { return "c-row"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["c-row.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-row.css"]
    };
  }
  static get properties() {
    return {
      "gap": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Gap between items in px"
        },
        "attribute": "gap",
        "reflect": false,
        "defaultValue": "0"
      },
      "nowrap": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Disable flex wrap"
        },
        "attribute": "nowrap",
        "reflect": false,
        "defaultValue": "false"
      },
      "align": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "CRowAlign",
          "resolved": "\"center\" | \"end\" | \"start\"",
          "references": {
            "CRowAlign": {
              "location": "import",
              "path": "../../types",
              "id": "src/types/index.ts::CRowAlign"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Align items vertically"
        },
        "attribute": "align",
        "reflect": false
      },
      "justify": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "CRowJustify",
          "resolved": "\"center\" | \"end\" | \"space-around\" | \"space-between\" | \"start\"",
          "references": {
            "CRowJustify": {
              "location": "import",
              "path": "../../types",
              "id": "src/types/index.ts::CRowJustify"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Justify content horizontally"
        },
        "attribute": "justify",
        "reflect": false
      }
    };
  }
  static get elementRef() { return "el"; }
}
