// @cscfi/csc-ui/dist/collection/components/c-list-item-title/c-list-item-title.js
import { Host, h } from "@stencil/core";
export class CListItemTitle {
  constructor() {
    this.active = false;
  }
  render() {
    const classes = {
      'c-list-item-title': true,
      'c-list-item-title--active': this.active,
    };
    return (h(Host, { class: classes }, h("slot", null)));
  }
  static get is() { return "c-list-item-title"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["c-list-item-title.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-list-item-title.css"]
    };
  }
  static get properties() {
    return {
      "active": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set the title active"
        },
        "attribute": "active",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
}
