// @cscfi/csc-ui/dist/collection/components/c-loader/c-loader.js
import { Host, h } from "@stencil/core";
export class CLoader {
  constructor() {
    this.contentdelay = 0;
    this.hide = false;
    this.size = 48;
    this.width = 4;
  }
  onElementHide(hide) {
    requestAnimationFrame(() => this.el.classList.toggle('active', !hide));
  }
  componentDidLoad() {
    requestAnimationFrame(() => this.el.classList.toggle('active', !this.hide));
  }
  render() {
    const slotHasContent = !!this.el.childNodes.length;
    const styles = {
      '--c-loader-size': `${this.size}px`,
    };
    return (h(Host, null, h("div", { class: "c-loader", style: styles }, h("c-spinner", { color: "var(--_c-loader-color)", size: this.size, width: this.width }), slotHasContent && (h("div", { class: "c-loader__slot", style: { 'animation-delay': `${this.contentdelay}s` } }, h("slot", null))))));
  }
  static get is() { return "c-loader"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["c-loader.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-loader.css"]
    };
  }
  static get properties() {
    return {
      "contentdelay": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Delay in seconds of showing the contents in the slot of the loader"
        },
        "attribute": "contentdelay",
        "reflect": false,
        "defaultValue": "0"
      },
      "hide": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Hide the loader"
        },
        "attribute": "hide",
        "reflect": false,
        "defaultValue": "false"
      },
      "size": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Size of the loader"
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "48"
      },
      "width": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Width of the loader"
        },
        "attribute": "width",
        "reflect": false,
        "defaultValue": "4"
      }
    };
  }
  static get elementRef() { return "el"; }
  static get watchers() {
    return [{
        "propName": "hide",
        "methodName": "onElementHide"
      }];
  }
}
