// @cscfi/csc-ui/dist/collection/components/c-side-navigation-item/c-side-navigation-item.js
import { h, Host, } from "@stencil/core";
import { mdiChevronRight } from "@mdi/js";
export class CSideNavigationItem {
  constructor() {
    this._isSubItem = false;
    this._slotHasContent = false;
    this.active = undefined;
    this.href = undefined;
    this.target = null;
    this.loading = false;
  }
  handleChange(event) {
    if (this._isSubItem) {
      event.stopPropagation();
      event.stopImmediatePropagation();
      event.preventDefault();
      this.active = !this.active;
    }
  }
  onActiveChange(active) {
    this._handleChildFocusableChange(active);
  }
  _handleChildFocusableChange(focusable) {
    if (!this._slotHasContent)
      return;
    Array.from(this.hostElement.querySelectorAll('[slot="sub-item"]')).forEach((child) => {
      child.ariaHidden = (!focusable).toString();
      child.focusable = focusable;
    });
  }
  _redirect(event) {
    if ((event instanceof KeyboardEvent && (event === null || event === void 0 ? void 0 : event.key) === 'Enter') ||
      !(event instanceof KeyboardEvent)) {
      if (this._isSubItem) {
        event.stopPropagation();
        event.stopImmediatePropagation();
        event.preventDefault();
      }
      this.itemChange.emit(event);
      if (!this._slotHasContent) {
        const sidenav = document.querySelector('c-side-navigation');
        sidenav.menuVisible = false;
      }
      if (this.href) {
        if (this.target) {
          window.open(this.href, this.target);
        }
        else {
          window.location.href = this.href;
        }
      }
    }
  }
  _assignSubItemSlots() {
    const subItems = this.hostElement.querySelectorAll('c-side-navigation-item');
    Array.from(subItems).forEach((item) => {
      item.setAttribute('slot', 'sub-item');
    });
  }
  componentWillLoad() {
    this._assignSubItemSlots();
    this._slotHasContent =
      !!this.hostElement.querySelector('[slot="sub-item"]');
    this._isSubItem = !!this.hostElement.getAttribute('slot');
    this._handleChildFocusableChange(this.active);
  }
  componentDidLoad() {
    for (const node of this.hostElement.childNodes) {
      if (node.nodeName === '#text') {
        this._ariaLabel = node.nodeValue.trim();
        break;
      }
    }
    this._handleChildClasses();
  }
  _handleChildClasses() {
    if (!this._isSubItem)
      return;
    Array.from(this.hostElement.children)
      .filter((child) => child.tagName === 'C-SUB-NAVIGATION-ITEM')
      .forEach((child) => {
      child.classList.add('c-sub-navigation-item--sub-level');
    });
  }
  render() {
    var _a, _b;
    const classes = {
      'c-side-navigation-item': true,
      'c-side-navigation-item--parent': this._slotHasContent,
      active: this.active,
    };
    const subNavigationClasses = {
      subnavactive: this.active,
      'sub-item': !this.active,
    };
    const a11y = {
      role: 'menuitem',
      tabindex: '0',
    };
    if (this._slotHasContent) {
      a11y['aria-expanded'] = (_a = (!!this.active)) === null || _a === void 0 ? void 0 : _a.toString();
    }
    else if (this.active) {
      a11y['aria-current'] = 'page';
    }
    return (h(Host, Object.assign({}, a11y, { class: classes, onClick: (e) => this._redirect(e), onKeyDown: (e) => this._redirect(e) }), h("div", { class: {
        'c-side-navigation-item__header': true,
        'c-side-navigation-item__header--expandable': this._slotHasContent,
      } }, this._slotHasContent && (h("c-icon", { class: "svg", path: mdiChevronRight })), h("div", { class: "c-side-navigation-item__slot" }, h("slot", null))), this._slotHasContent && (h("nav", { role: "menubar", "aria-label": this._ariaLabel, "aria-expanded": (_b = (!!this.active)) === null || _b === void 0 ? void 0 : _b.toString(), class: subNavigationClasses }, h("slot", { name: "sub-item" }))), h("c-loader", { size: 32, hide: !this.loading, style: { pointerEvents: 'none' } })));
  }
  static get is() { return "c-side-navigation-item"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["c-side-navigation-item.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-side-navigation-item.css"]
    };
  }
  static get properties() {
    return {
      "active": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Indicate active state"
        },
        "attribute": "active",
        "reflect": false
      },
      "href": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Hyperlink url"
        },
        "attribute": "href",
        "reflect": false
      },
      "target": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Hyperlink target"
        },
        "attribute": "target",
        "reflect": false,
        "defaultValue": "null"
      },
      "loading": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Loading state"
        },
        "attribute": "loading",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
  static get events() {
    return [{
        "method": "itemChange",
        "name": "itemChange",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [{
              "name": "private",
              "text": undefined
            }],
          "text": "Emit changes to the c-accordion"
        },
        "complexType": {
          "original": "any",
          "resolved": "any",
          "references": {}
        }
      }];
  }
  static get elementRef() { return "hostElement"; }
  static get watchers() {
    return [{
        "propName": "active",
        "methodName": "onActiveChange"
      }];
  }
  static get listeners() {
    return [{
        "name": "itemChange",
        "method": "handleChange",
        "target": undefined,
        "capture": false,
        "passive": false
      }];
  }
}
