// @cscfi/csc-ui/dist/collection/components/c-switch/c-switch.js
import { h, } from "@stencil/core";
export class CSwitch {
  constructor() {
    this._toggleState = () => {
      if (this.hostDisabled)
        return;
      this.checked = !this.checked;
      this.changeValue.emit(this.checked ? this.trueValue : this.falseValue);
      this.internals.setFormValue(this.checked ? this.trueValue.toString() : this.falseValue.toString());
    };
    this.checked = false;
    this.value = false;
    this.trueValue = true;
    this.falseValue = false;
    this.hostDisabled = false;
    this.hostId = undefined;
    this.required = false;
    this.hasLabel = false;
  }
  componentWillLoad() {
    this.checked = this.checked || this.value === this.trueValue;
    this.internals.setFormValue(this.checked ? this.trueValue.toString() : this.falseValue.toString());
  }
  componentDidLoad() {
    const slotted = this.host.childNodes;
    this.hasLabel = slotted && slotted.length > 0;
  }
  render() {
    const classes = {
      'c-switch': true,
      'c-switch--disabled': !!this.hostDisabled,
      'c-switch--label': this.hasLabel,
    };
    const sliderClasses = {
      'c-switch__slider': true,
      'c-switch__slider--disabled': !!this.hostDisabled,
    };
    return (h("label", { class: classes, htmlFor: this.hostId }, h("div", { class: "c-switch__input" }, h("input", { id: this.hostId, "aria-checked": this.checked.toString(), type: "checkbox", role: "switch", disabled: this.hostDisabled, checked: this.checked ? true : undefined, onChange: () => this._toggleState() }), h("span", { class: sliderClasses })), this.hasLabel ? (h("div", { class: "c-switch__label" }, h("slot", null), this.required && h("span", { class: "required" }, "\u00A0*"))) : null));
  }
  static get is() { return "c-switch"; }
  static get encapsulation() { return "shadow"; }
  static get formAssociated() { return true; }
  static get originalStyleUrls() {
    return {
      "$": ["c-switch.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-switch.css"]
    };
  }
  static get properties() {
    return {
      "checked": {
        "type": "boolean",
        "mutable": true,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "If `true`, the checkbox is selected."
        },
        "attribute": "checked",
        "reflect": false,
        "defaultValue": "false"
      },
      "value": {
        "type": "any",
        "mutable": false,
        "complexType": {
          "original": "string | boolean",
          "resolved": "boolean | string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The input value\n- Only used when the checkbox participates in a native `<form>`"
        },
        "attribute": "value",
        "reflect": false,
        "defaultValue": "false"
      },
      "trueValue": {
        "type": "any",
        "mutable": false,
        "complexType": {
          "original": "boolean | string",
          "resolved": "boolean | string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The value when the checkbox is checked"
        },
        "attribute": "true-value",
        "reflect": false,
        "defaultValue": "true"
      },
      "falseValue": {
        "type": "any",
        "mutable": false,
        "complexType": {
          "original": "boolean | string",
          "resolved": "boolean | string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The value when the checkbox is unchecked"
        },
        "attribute": "false-value",
        "reflect": false,
        "defaultValue": "false"
      },
      "hostDisabled": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Disable the switch"
        },
        "attribute": "disabled",
        "reflect": false,
        "defaultValue": "false"
      },
      "hostId": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Id for the element"
        },
        "attribute": "id",
        "reflect": false
      },
      "required": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Set as required"
        },
        "attribute": "required",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
  static get states() {
    return {
      "hasLabel": {}
    };
  }
  static get events() {
    return [{
        "method": "changeValue",
        "name": "changeValue",
        "bubbles": false,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [],
          "text": "Emit inner value change to parent"
        },
        "complexType": {
          "original": "any",
          "resolved": "any",
          "references": {}
        }
      }];
  }
  static get elementRef() { return "host"; }
  static get attachInternalsMemberName() { return "internals"; }
}
