// @cscfi/csc-ui/dist/collection/components/c-tags/c-tags.js
import { Host, h } from "@stencil/core";
export class CTags {
  constructor() {
    this.size = 'default';
  }
  componentDidLoad() {
    if (this.size === 'small') {
      const tags = this.el.querySelectorAll('c-tag');
      [...tags].forEach((tag) => {
        tag.size = 'small';
      });
    }
  }
  render() {
    return (h(Host, null, h("slot", null)));
  }
  static get is() { return "c-tags"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["c-tags.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-tags.css"]
    };
  }
  static get properties() {
    return {
      "size": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "'default' | 'small'",
          "resolved": "\"default\" | \"small\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Size of the tags"
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "'default'"
      }
    };
  }
  static get elementRef() { return "el"; }
}
