// @cscfi/csc-ui/dist/cjs/c-accordion.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-91ba22ab.js');

const cAccordionCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}.c-accordion{display:flex;flex-wrap:wrap;gap:8px}.c-accordion ::slotted(*){flex-basis:100%}";

const CAccordion = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.changeValue = index.createEvent(this, "changeValue", 3);
    this.value = undefined;
    this.mandatory = false;
    this.multiple = false;
    this.outlined = false;
  }
  watchPropHandler(value) {
    this.changeValue.emit(value);
    this._handleItemExpansion();
  }
  onAccordionItemChange(event) {
    const { value, expanded } = event.detail;
    if (this.multiple && Array.isArray(this.value)) {
      if (expanded) {
        this.value.push(value);
      }
      else {
        this.value = this.value.filter((v) => v !== value);
      }
    }
    else {
      this.value = expanded ? value : null;
    }
  }
  get items() {
    return Array.from(this.el.childNodes);
  }
  _handleItemExpansion() {
    const isExpanded = (item) => Array.isArray(this.value)
      ? this.value.includes(item.value)
      : item.value === this.value;
    const isLastExpandedItem = (item) => Array.isArray(this.value)
      ? isExpanded(item) && this.value.length === 1
      : isExpanded(item);
    for (const item of this.items) {
      item.collapsable =
        !this.mandatory || (this.mandatory && !isLastExpandedItem(item));
      item.outlined = this.outlined;
      item.expanded = isExpanded(item);
    }
  }
  componentDidLoad() {
    this._handleItemExpansion();
  }
  render() {
    return (index.h("div", { class: "c-accordion" }, index.h("slot", null)));
  }
  get el() { return index.getElement(this); }
  static get watchers() { return {
    "value": ["watchPropHandler"]
  }; }
};
CAccordion.style = cAccordionCss;

exports.c_accordion = CAccordion;
