// @cscfi/csc-ui/dist/cjs/c-alert.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-91ba22ab.js');
const mdi = require('./mdi-66dcba4d.js');

const cAlertCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}.c-alert{--c-alert-color:var(--c-primary-600);display:grid;gap:16px;grid-template-columns:1fr;color:var(--c-alert-color);border:2px solid currentColor;border-left-width:12px;border-radius:6px;padding:12px}.c-alert--info,.c-alert--error,.c-alert--success,.c-alert--warning{grid-template-columns:auto 1fr}.c-alert--info{--c-alert-color:var(--c-info-600)}.c-alert--error{--c-alert-color:var(--c-error-600)}.c-alert--success{--c-alert-color:var(--c-success-600)}.c-alert--warning{--c-alert-color:var(--c-warning-600)}.c-alert__content{color:rgba(0, 0, 0, 0.87);display:grid;align-items:center;grid-template-columns:1fr;gap:8px}svg{fill:currentColor}::slotted(*[slot=title]){margin:0 !important;font-size:18px;font-weight:600;line-height:24px}";

const CAlert = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this._icons = {
      warning: mdi.mdiAlert,
      error: mdi.mdiCloseCircle,
      success: mdi.mdiCheckCircle,
      info: mdi.mdiInformation,
    };
    this.type = undefined;
  }
  render() {
    const classes = Object.assign({ 'c-alert': true }, (!!this.type && { [`c-alert--${this.type}`]: true }));
    return (index.h(index.Host, null, index.h("div", { class: classes }, !!this.type && (index.h("svg", { width: "22", height: "22", viewBox: "0 0 24 24" }, index.h("path", { d: this._icons[this.type] }))), index.h("div", { class: "c-alert__content" }, index.h("slot", { name: "title" }), index.h("slot", null)))));
  }
};
CAlert.style = cAlertCss;

exports.c_alert = CAlert;
