// @cscfi/csc-ui/dist/cjs/c-row.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-91ba22ab.js');

const cRowCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host(.c-row){display:flex;flex:1 1 auto;flex-wrap:wrap;gap:var(--_c-row-gap)}:host(.c-row) slot{display:flex;flex-wrap:wrap;width:100%;gap:var(--_c-row-gap)}:host(.c-row--nowrap) slot{flex-wrap:nowrap !important}:host(.c-row--align-center) slot{align-items:center}:host(.c-row--align-start) slot{align-items:flex-start}:host(.c-row--align-end) slot{align-items:flex-end}:host(.c-row--justify-start) slot{justify-content:flex-start}:host(.c-row--justify-center) slot{justify-content:center}:host(.c-row--justify-end) slot{justify-content:flex-end}:host(.c-row--justify-space-between) slot{justify-content:space-between}:host(.c-row--justify-space-around) slot{justify-content:space-around}";

const CRow = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.gap = 0;
    this.nowrap = false;
    this.align = undefined;
    this.justify = undefined;
  }
  componentDidLoad() {
    this.el.style.setProperty('--_c-row-gap', this.el.dataset.gap);
  }
  render() {
    const classes = {
      'c-row': true,
      'c-row--nowrap': this.nowrap,
      [`c-row--align-${this.align}`]: !!this.align,
      [`c-row--justify-${this.justify}`]: !!this.justify,
    };
    return (index.h(index.Host, { "data-gap": `${this.gap}px`, class: classes }, index.h("slot", null)));
  }
  get el() { return index.getElement(this); }
};
CRow.style = cRowCss;

exports.c_row = CRow;
