// @cscfi/csc-ui/dist/cjs/c-tab-item.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-91ba22ab.js');

const cTabItemCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host{--_c-tab-item-padding:var(--c-tab-item-padding, 16px 0 0 0);display:block;min-width:100%;padding:var(--_c-tab-item-padding);height:0;user-select:none}:host([active=true]){height:auto;user-select:auto}:host([active=false])>div{display:none}:host([disabled=true]){min-width:0;width:0;overflow:hidden}";

const CTabItem = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.contentChange = index.createEvent(this, "contentChange", 7);
    this.value = undefined;
    this.active = false;
  }
  componentDidLoad() {
    this._resizeObserver = new ResizeObserver(() => {
      this._handleResize();
    });
    this._resizeObserver.observe(this.el.shadowRoot.querySelector('div'));
  }
  disconnectedCallback() {
    this._resizeObserver.disconnect();
  }
  _handleResize() {
    this.contentChange.emit();
  }
  render() {
    const attributes = {
      role: 'tabpanel',
    };
    return (index.h(index.Host, Object.assign({}, attributes), index.h("div", { class: { active: this.active } }, index.h("slot", null))));
  }
  get el() { return index.getElement(this); }
};
CTabItem.style = cTabItemCss;

exports.c_tab_item = CTabItem;
