// @cscfi/csc-ui/dist/components/c-accordion.js
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';

const cAccordionCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}.c-accordion{display:flex;flex-wrap:wrap;gap:8px}.c-accordion ::slotted(*){flex-basis:100%}";

const CAccordion$1 = proxyCustomElement(class CAccordion extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.changeValue = createEvent(this, "changeValue", 3);
    this.value = undefined;
    this.mandatory = false;
    this.multiple = false;
    this.outlined = false;
  }
  watchPropHandler(value) {
    this.changeValue.emit(value);
    this._handleItemExpansion();
  }
  onAccordionItemChange(event) {
    const { value, expanded } = event.detail;
    if (this.multiple && Array.isArray(this.value)) {
      if (expanded) {
        this.value.push(value);
      }
      else {
        this.value = this.value.filter((v) => v !== value);
      }
    }
    else {
      this.value = expanded ? value : null;
    }
  }
  get items() {
    return Array.from(this.el.childNodes);
  }
  _handleItemExpansion() {
    const isExpanded = (item) => Array.isArray(this.value)
      ? this.value.includes(item.value)
      : item.value === this.value;
    const isLastExpandedItem = (item) => Array.isArray(this.value)
      ? isExpanded(item) && this.value.length === 1
      : isExpanded(item);
    for (const item of this.items) {
      item.collapsable =
        !this.mandatory || (this.mandatory && !isLastExpandedItem(item));
      item.outlined = this.outlined;
      item.expanded = isExpanded(item);
    }
  }
  componentDidLoad() {
    this._handleItemExpansion();
  }
  render() {
    return (h("div", { class: "c-accordion" }, h("slot", null)));
  }
  get el() { return this; }
  static get watchers() { return {
    "value": ["watchPropHandler"]
  }; }
  static get style() { return cAccordionCss; }
}, [1, "c-accordion", {
    "value": [1032],
    "mandatory": [4],
    "multiple": [4],
    "outlined": [4]
  }, [[1, "itemChange", "onAccordionItemChange"]], {
    "value": ["watchPropHandler"]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["c-accordion"];
  components.forEach(tagName => { switch (tagName) {
    case "c-accordion":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, CAccordion$1);
      }
      break;
  } });
}

const CAccordion = CAccordion$1;
const defineCustomElement = defineCustomElement$1;

export { CAccordion, defineCustomElement };
