// @cscfi/csc-ui/dist/components/c-autocomplete.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { e as mdiChevronDown, f as mdiClose } from './mdi.js';
import { d as defineCustomElement$9 } from './c-badge2.js';
import { d as defineCustomElement$8 } from './c-dropdown2.js';
import { d as defineCustomElement$7 } from './c-icon2.js';
import { d as defineCustomElement$6 } from './c-icon-button2.js';
import { d as defineCustomElement$5 } from './c-input2.js';
import { d as defineCustomElement$4 } from './c-message2.js';
import { d as defineCustomElement$3 } from './c-ripple2.js';
import { d as defineCustomElement$2 } from './c-spinner2.js';

const cAutocompleteCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host(c-autocomplete){--_c-autocomplete-option-background-color:var(\n    --c-autocomplete-option-background-color,\n    var(--c-white)\n  );--_c-autocomplete-option-background-hover-color:var(\n    --c-autocomplete-option-background-hover-color,\n    var(--c-primary-100)\n  );--_c-autocomplete-active-color:var(\n    --c-autocomplete-active-color,\n    var(--c-primary-600)\n  );--_c-autocomplete-inactive-color:var(\n    --c-autocomplete-inactive-color,\n    var(--c-tertiary-600)\n  );--_c-autocomplete-background-color:var(\n    --c-autocomplete-background-color,\n    var(--c-transparent)\n  );--_c-autocomplete-text-color:var(\n    --c-autocomplete-text-color,\n    var(--c-text-body)\n  );--_c-autocomplete-option-text-color:var(\n    --c-autocomplete-option-text-color,\n    var(--c-text-body)\n  );--_c-autocomplete-placeholder-color:var(\n    --c-autocomplete-placeholder-color,\n    var(--c-tertiary-500)\n  );--c-select-option-background-color:var(\n    --_c-autocomplete-option-background-color\n  );--c-select-option-background-color-hover:var(\n    --_c-autocomplete-option-background-hover-color\n  );--c-select-option-text-color:var(--_c-autocomplete-option-text-color);--_c-input-active-color:var(--_c-autocomplete-active-color);--_c-input-inactive-color:var(--_c-autocomplete-inactive-color);--_c-input-background-color:var(--_c-autocomplete-background-color);--_c-input-text-color:var(--_c-autocomplete-text-color);--_c-input-placeholder-color:var(--_c-autocomplete-placeholder-color)}:host{display:block;cursor:text}.c-input-menu__chevron{fill:currentColor;transform:rotate(0deg);transition:transform 0.3s ease-in-out}.c-input-menu__chevron--active{transform:rotate(180deg)}.c-input-menu__input{width:100%;display:flex;justify-items:stretch}.c-input-menu__selection{display:none;pointer-events:none}.c-input-menu__selection--show{align-items:center;display:flex;width:100%;color:var(--_c-select-active-color)}.c-input-menu__item-wrapper{position:absolute;width:100%;top:44px;z-index:10;margin-left:calc(var(--_c-input-label-position) * -1)}.c-input-menu__item-wrapper--shadow{top:47px}.c-input-menu__items{position:absolute;background-color:var(--_c-select-option-background-color);min-width:calc(100% + 24px);box-shadow:0 8px 16px 0 rgba(0, 0, 0, 0.2);z-index:10;user-select:none;border-radius:4px;margin:0 -12px;overflow-y:scroll;list-style:none;padding:0}.c-input-menu__items--hidden{display:none}.c-input-menu__items li{cursor:pointer;display:flex;min-height:48px;padding:0 12px;transition:background-color 0.3s;font-size:14px;align-items:center;justify-content:flex-start;color:var(--_c-select-option-text-color)}.c-input-menu__items li.disabled{background-color:var(--c-tertiary-100);filter:grayscale(1) opacity(0.6);cursor:default}.c-input-menu__items li.disabled:hover{background-color:var(--c-tertiary-100)}.c-input-menu__items li.dense{padding:10px 14px}.c-input-menu__items li:hover{background-color:var(--_c-select-option-background-color-hover)}.c-input-menu__items li[aria-selected=true]{background-color:var(--_c-select-option-background-color-hover);box-shadow:inset 0 0 0 1px var(--_c-select-active-color);border-radius:4px;backface-visibility:hidden;transform:translate3d(0, 0, 0)}.c-input-menu__items li.none{color:rgba(0, 0, 0, 0.5)}.c-input-menu__items--empty li{color:var(--c-text-system);cursor:default;gap:8px;pointer-events:none}.c-input-menu__items--empty li svg{fill:var(--c-warning-600);height:18px;width:18px}input{max-height:32px;padding:8px 0;background-color:transparent;border:none;color:var(--c-text-body);flex:1 1 auto;font-family:var(--c-font-family);font-size:16px;line-height:20px;max-width:100%;min-width:0;width:100%;pointer-events:none}input:focus,input:active{outline:none}input::-ms-reveal{display:none}svg{fill:currentColor;height:22px;width:22px}.c-input--disabled{color:var(--c-tertiary-400)}.c-input:focus-within{color:var(--c-primary-600)}.c-input--error{color:var(--c-error-600)}.c-input--error:focus-within{color:var(--c-error-600)}.visuallyhidden{border:0;clip:rect(0 0 0 0);height:1px;margin:-1px;overflow:hidden;padding:0;position:absolute;white-space:nowrap;width:1px}";

const CAutocomplete$1 = proxyCustomElement(class CAutocomplete extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.changeQuery = createEvent(this, "changeQuery", 7);
    this.changeValue = createEvent(this, "changeValue", 3);
    this.internals = this.attachInternals();
    this._preventDialogOpen = false;
    this._debounce = null;
    this._handleSlotChange = () => {
      this.optionElements = this.el.querySelectorAll('c-option');
      const selection = Array.from(this.optionElements).find((option) => option.selected);
      if (selection) {
        this.value = this.returnObject
          ? { name: selection.name, value: selection.value }
          : selection.value;
        this.changeValue.emit(this.value);
        this.changeQuery.emit(this.query);
        this.internals.setFormValue(selection.value.toString());
      }
      if (this.optionElements.length && !this.optionElementsExist) {
        this.optionElementsExist = true;
      }
    };
    this._toggleDropdown = (event) => {
      event.stopPropagation();
      if (this.dropdownVisible) {
        this._dropdownElement.close();
        return;
      }
      this._dropdownElement.open();
    };
    this._onButtonKeyDown = (src, event) => {
      event.stopPropagation();
      if (event.key !== 'Tab')
        event.preventDefault();
      if (['Enter', ' '].includes(event.key)) {
        if (src === 'chevron') {
          this._toggleDropdown(event);
          return;
        }
        this._reset(event);
        this._preventDialogOpen = true;
        this._inputElement.focus();
      }
    };
    this._reset = (event) => {
      event.stopPropagation();
      this.query = '';
      this.value = null;
      this.changeValue.emit(null);
      this.internals.setFormValue(null);
      this.changeQuery.emit('');
      this._preventDialogOpen = true;
      this._cInputElement.reset();
      this._inputElement.focus();
      this._dropdownElement.updateList();
    };
    this._onInputFocus = () => {
      if (!this._preventDialogOpen) {
        this._dropdownElement.open();
      }
      this._updateStatusText();
      this._preventDialogOpen = false;
    };
    this.items = [];
    this.value = null;
    this.query = null;
    this.hostId = undefined;
    this.disabled = false;
    this.label = undefined;
    this.name = undefined;
    this.placeholder = '';
    this.hideDetails = false;
    this.hint = '';
    this.loading = false;
    this.required = false;
    this.valid = true;
    this.validate = false;
    this.validateOnBlur = false;
    this.validation = 'Required field';
    this.shadow = false;
    this.returnObject = false;
    this.itemsPerPage = 6;
    this.minimumQueryLength = 0;
    this.minimumQueryLengthMessage = 'Type at least {n} characters';
    this.noMatchingItemsMessage = 'No suggestions found';
    this.optionElements = undefined;
    this.currentIndex = null;
    this.dropdownVisible = false;
    this.statusText = '';
    this.optionElementsExist = false;
  }
  get _id() {
    return this.hostId || `autocomplete_${CAutocomplete._uniqueId}`;
  }
  async reset() {
    this.query = '';
    this.changeValue.emit(null);
    this.internals.setFormValue(null);
    this.changeQuery.emit('');
    this._dropdownElement.updateList();
  }
  _handleKeyDown(event) {
    const alphanumeric = /^[0-9a-zA-Z ]+$/;
    if (event.key.match(alphanumeric) && event.key.length === 1) {
      if (!this.dropdownVisible) {
        this._dropdownElement.open();
        return;
      }
      this._inputElement.focus();
      return;
    }
    if (event.key === 'Escape') {
      event.preventDefault();
      this._preventDialogOpen = true;
      this._dropdownElement.close();
      requestAnimationFrame(() => {
        this._inputElement.focus();
      });
      return;
    }
    if (event.key === 'Tab') {
      this._inputElement.focus();
      requestAnimationFrame(() => {
        this._dropdownElement.close();
      });
      return;
    }
    if (event.key === 'ArrowDown') {
      event.preventDefault();
      if (!this._items.length)
        return;
      if (!this.dropdownVisible || this.currentIndex === null) {
        this.currentIndex = 0;
        this._dropdownElement.open();
        this._dropdownElement.focusItem(this.currentIndex);
        return;
      }
      this.currentIndex = Math.min(this.currentIndex + 1, this._items.length - 1);
      this._dropdownElement.focusItem(this.currentIndex);
      return;
    }
    if (event.key === 'ArrowUp') {
      event.preventDefault();
      if (this.currentIndex === 0) {
        this._dropdownElement.close();
        this._inputElement.focus();
      }
      if (!this.dropdownVisible || this.currentIndex === null) {
        this.currentIndex = this._items.length - 1;
        this._dropdownElement.open();
        this._dropdownElement.focusItem(this.currentIndex);
        return;
      }
      this.currentIndex = Math.max(this.currentIndex - 1, 0);
      this._dropdownElement.focusItem(this.currentIndex);
    }
    if (event.key === 'Enter') {
      event.preventDefault();
      if (this.currentIndex === null)
        return;
      this._dropdownElement.selectItem(this.currentIndex);
    }
    if (event.key === 'Home' && this.dropdownVisible) {
      this.currentIndex = 0;
    }
    if (event.key === 'End' && this.dropdownVisible) {
      this.currentIndex = this._items.length - 1;
    }
  }
  onDropdownStateChange(event) {
    const isOpen = event.detail;
    this.dropdownVisible = isOpen;
    if (!isOpen) {
      requestAnimationFrame(() => {
        this.currentIndex = null;
      });
    }
  }
  onSelectOption(event) {
    this._dropdownElement.close();
    const { name, value } = event.detail;
    this.value = this.returnObject ? event.detail : value;
    this.changeValue.emit(this.value);
    this.internals.setFormValue(value);
    this.query = name;
    this.changeQuery.emit(name);
    this._preventDialogOpen = true;
    if (this.optionElementsExist) {
      Array.from(this.optionElements).forEach((item) => {
        item.selected = item.value === value && item.name === name;
      });
      this._dropdownElement.updateList();
    }
    this._inputElement.focus();
  }
  handleQueryChange() {
    this._updateStatusText();
  }
  _updateInput() {
    this._dropdownElement.open();
    this.query = this._inputElement.value;
    this.changeQuery.emit(this.query);
    this._dropdownElement.updateList();
  }
  _renderLoader() {
    return h("c-spinner", { color: "var(--_c-autocomplete-active-color)", size: 20 });
  }
  _renderChevron() {
    const classes = {
      'c-input-menu__chevron': true,
      'c-input-menu__chevron--active': this.dropdownVisible,
    };
    return (h("c-icon-button", { size: "x-small", class: classes, disabled: this.disabled, text: true, onClick: (event) => this._toggleDropdown(event), onKeyDown: (event) => this._onButtonKeyDown('chevron', event) }, h("c-icon", { path: mdiChevronDown, size: 24 })));
  }
  _renderReset() {
    return (h("c-icon-button", { "aria-label": "", size: "x-small", disabled: this.disabled, text: true, onClick: (event) => this._reset(event), onKeyDown: (event) => this._onButtonKeyDown('reset', event) }, h("c-icon", { path: mdiClose, size: 20 })));
  }
  _renderInputElement() {
    var _a;
    return (h("div", { class: "c-input-menu__input" }, h("input", { type: "text", ref: (el) => (this._inputElement = el), "aria-expanded": this.dropdownVisible.toString(), "aria-owns": this._inputId + '-items', "aria-autocomplete": "list", autocomplete: "off", class: "c-input__input", role: "combobox", value: this.query, name: (_a = this.name) !== null && _a !== void 0 ? _a : null, disabled: this.disabled, onInput: () => this._updateInput(), onFocus: () => this._onInputFocus() })));
  }
  componentWillLoad() {
    CAutocomplete._uniqueId += 1;
    this._inputId =
      'input_' +
        (this.hostId || this.label || this.placeholder).replace(/[^a-zA-Z0-9-_]/g, '');
  }
  get _items() {
    return this.optionElementsExist ? this.optionElements : this.items;
  }
  _updateStatusText() {
    if (this._debounce !== null) {
      clearTimeout(this._debounce);
      this._debounce = null;
    }
    this._debounce = window.setTimeout(() => {
      const ending = !!this._items.length
        ? ' or navigate using the up and down arrows'
        : '';
      if (this.currentIndex === null) {
        this.statusText = this._items.length
          ? `${this._items.length} result${this._items.length !== 1 ? 's' : ''} available`
          : 'No search results available';
      }
      if (this.query !== null && this._items.length) {
        this.statusText += ', input a search query to filter the results';
      }
      this._dropdownElement.setStatusText(this.statusText + ending);
      this._debounce = null;
    }, 1400);
  }
  render() {
    const itemType = this.optionElementsExist ? 'option' : 'item';
    return (h(Host, null, h("c-dropdown", { ref: (el) => (this._dropdownElement = el), id: `${this._id}-dropdown`, index: this.currentIndex, "items-per-page": this.itemsPerPage, "item-type": itemType, items: this._items, parent: this.el, type: "autocomplete", onKeyDown: (event) => this._handleKeyDown(event) }, h("c-input", { slot: "default", ref: (el) => (this._cInputElement = el), active: this.dropdownVisible, disabled: this.disabled, "hide-details": this.hideDetails, hint: this.hint, id: this.hostId, "input-id": this._inputId, label: this.label, name: this.name, placeholder: this.placeholder, required: this.required, shadow: this.shadow, valid: this.valid, validate: this.validate, "validate-on-blur": this.validateOnBlur, validation: this.validation, value: this.query, variant: "select" }, h("slot", { name: "pre", slot: "pre" }), h("div", { class: "c-input__content" }, this._renderInputElement(), this.loading && this._renderLoader(), !this.loading && this.value && this._renderReset(), !this.loading && !this.value && this._renderChevron(), h("slot", { onSlotchange: () => this._handleSlotChange() })), h("slot", { name: "post", slot: "post" })))));
  }
  static get formAssociated() { return true; }
  get el() { return this; }
  static get watchers() { return {
    "query": ["handleQueryChange"]
  }; }
  static get style() { return cAutocompleteCss; }
}, [65, "c-autocomplete", {
    "items": [16],
    "value": [8],
    "query": [1],
    "hostId": [1, "id"],
    "disabled": [4],
    "label": [1],
    "name": [1],
    "placeholder": [1],
    "hideDetails": [4, "hide-details"],
    "hint": [1],
    "loading": [4],
    "required": [4],
    "valid": [4],
    "validate": [4],
    "validateOnBlur": [4, "validate-on-blur"],
    "validation": [1],
    "shadow": [4],
    "returnObject": [4, "return-object"],
    "itemsPerPage": [2, "items-per-page"],
    "minimumQueryLength": [2, "minimum-query-length"],
    "minimumQueryLengthMessage": [1, "minimum-query-length-message"],
    "noMatchingItemsMessage": [1, "no-matching-items-message"],
    "optionElements": [32],
    "currentIndex": [32],
    "dropdownVisible": [32],
    "statusText": [32],
    "optionElementsExist": [32],
    "reset": [64]
  }, [[0, "dropdownStateChange", "onDropdownStateChange"], [0, "selectOption", "onSelectOption"]], {
    "query": ["handleQueryChange"]
  }]);
CAutocomplete$1._uniqueId = 0;
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["c-autocomplete", "c-badge", "c-dropdown", "c-icon", "c-icon-button", "c-input", "c-message", "c-ripple", "c-spinner"];
  components.forEach(tagName => { switch (tagName) {
    case "c-autocomplete":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, CAutocomplete$1);
      }
      break;
    case "c-badge":
      if (!customElements.get(tagName)) {
        defineCustomElement$9();
      }
      break;
    case "c-dropdown":
      if (!customElements.get(tagName)) {
        defineCustomElement$8();
      }
      break;
    case "c-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$7();
      }
      break;
    case "c-icon-button":
      if (!customElements.get(tagName)) {
        defineCustomElement$6();
      }
      break;
    case "c-input":
      if (!customElements.get(tagName)) {
        defineCustomElement$5();
      }
      break;
    case "c-message":
      if (!customElements.get(tagName)) {
        defineCustomElement$4();
      }
      break;
    case "c-ripple":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "c-spinner":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const CAutocomplete = CAutocomplete$1;
const defineCustomElement = defineCustomElement$1;

export { CAutocomplete, defineCustomElement };
