// @cscfi/csc-ui/dist/components/c-card-actions2.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const cCardActionsCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host{padding:0 var(--_c-card-gap, 24px);display:block}:host menu{margin:0;padding:0}.c-card-actions{align-items:center;display:flex;gap:8px}.c-card-actions ::slotted(c-button[fit]){flex:1}.c-card-actions--align-center{align-items:center}.c-card-actions--align-start{align-items:flex-start}.c-card-actions--align-end{align-items:flex-end}.c-card-actions--justify-center{justify-content:center}.c-card-actions--justify-end{justify-content:flex-end}.c-card-actions--justify-start{justify-content:flex-start}.c-card-actions--justify-stretch{justify-content:stretch}.c-card-actions--justify-stretch ::slotted(*){flex:1}.c-card-actions--justify-space-between{justify-content:space-between}.c-card-actions--justify-space-around{justify-content:space-around}";

const CCardActions = proxyCustomElement(class CCardActions extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.align = 'center';
    this.justify = 'start';
  }
  render() {
    const classes = {
      'c-card-actions': true,
      [`c-card-actions--align-${this.align}`]: true,
      [`c-card-actions--justify-${this.justify}`]: true,
    };
    return (h("footer", null, h("menu", { class: classes }, h("slot", null))));
  }
  static get style() { return cCardActionsCss; }
}, [1, "c-card-actions", {
    "align": [1],
    "justify": [1]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["c-card-actions"];
  components.forEach(tagName => { switch (tagName) {
    case "c-card-actions":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, CCardActions);
      }
      break;
  } });
}

export { CCardActions as C, defineCustomElement as d };
