// @cscfi/csc-ui/dist/components/c-radio.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';

const cRadioCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host{display:block}";

const CRadio$1 = proxyCustomElement(class CRadio extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.checked = false;
    this.disabled = false;
    this.value = undefined;
  }
  render() {
    return h(Host, null);
  }
  static get style() { return cRadioCss; }
}, [1, "c-radio", {
    "checked": [4],
    "disabled": [4],
    "value": [1]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["c-radio"];
  components.forEach(tagName => { switch (tagName) {
    case "c-radio":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, CRadio$1);
      }
      break;
  } });
}

const CRadio = CRadio$1;
const defineCustomElement = defineCustomElement$1;

export { CRadio, defineCustomElement };
