// @cscfi/csc-ui/dist/components/c-steps.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';

const cStepsCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host{--_c-steps-outline-color:var(--c-steps-outline-color, var(--c-primary-600));--_c-steps-color:var(--c-step-color, var(--c-tertiary-500));--_c-steps-color-complete:var(--c-step-color-complete, var(--c-primary-600));width:100%;display:grid;gap:8px;border-radius:6px}@supports selector(:focus-visible){:host(:focus){outline:none}}:host(:focus-visible){outline:2px var(--_c-steps-outline-color) solid;outline-offset:2px}.c-steps{--c-steps-divider-width:calc(100% + 160px);--c-steps-divider-margin:10px -80px 0;width:100%;display:flex;flex-wrap:nowrap}.c-steps slot{display:flex;justify-items:space-between;width:100%}.c-steps ::slotted(.divider){height:2px;flex:1;background-color:var(--_c-steps-color);width:var(--c-steps-divider-width);margin:var(--c-steps-divider-margin)}.c-steps ::slotted(.divider.complete){background-color:var(--_c-steps-color-complete);height:4px;margin-top:9px}.c-steps.mobile{--c-steps-divider-width:calc(100% + 11px);--c-steps-divider-margin:10px -10px 0}.visuallyhidden{border:0;clip:rect(0 0 0 0);height:1px;margin:-1px;overflow:hidden;padding:0;position:absolute;white-space:nowrap;width:1px}";

const CSteps$1 = proxyCustomElement(class CSteps extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.changeValue = createEvent(this, "changeValue", 3);
    this._isInitialized = false;
    this.value = undefined;
    this.isMobile = false;
    this.label = '';
  }
  watchPropHandler() {
    this._handleDividers();
  }
  _handleDividers() {
    this._stepElements = this.host.querySelectorAll('c-step');
    const dividers = this.host.querySelectorAll('.divider');
    this._stepElements.forEach((item, index) => {
      item.current = index + 1 === +this.value;
      item.complete = index + 1 < +this.value;
      if (index + 1 < this._stepElements.length) {
        const div = this._isInitialized
          ? dividers[index]
          : document.createElement('div');
        div.classList.toggle('complete', !!item.complete);
        if (!this._isInitialized) {
          div.classList.add('divider');
          item.after(div);
        }
      }
      if (item.current) {
        this.label = item.textContent;
      }
    });
    this._isInitialized = true;
  }
  componentDidLoad() {
    this._handleDividers();
    this._resizeObserver = new ResizeObserver(([entry]) => {
      const maxWidth = this._stepElements.length * 180;
      this.isMobile = maxWidth > entry.contentRect.width;
      this.host.shadowRoot
        .querySelector('.c-steps')
        .classList.toggle('mobile', this.isMobile);
      this._stepElements.forEach((node) => {
        node.classList.toggle('mobile', this.isMobile);
      });
    });
    window.requestAnimationFrame(() => {
      this._resizeObserver.observe(this.host);
    });
  }
  _getA11yMessage(total, current) {
    return `
      Steps, step ${Math.min(current, total)} of ${total}.
      ${this.label}.
      ${current - 1} step${current - 1 !== 1 ? 's' : ''} marked as completed.
    `;
  }
  render() {
    return (h(Host, { tabindex: "0" }, this._stepElements && (h("span", { class: "visuallyhidden" }, this._getA11yMessage(this._stepElements.length, +this.value))), h("div", { class: "c-steps", "aria-hidden": "true" }, h("slot", null)), this.isMobile && (h("div", { class: "c-steps__label", "aria-hidden": "true" }, this.label))));
  }
  get host() { return this; }
  static get watchers() { return {
    "value": ["watchPropHandler"]
  }; }
  static get style() { return cStepsCss; }
}, [1, "c-steps", {
    "value": [8],
    "isMobile": [32],
    "label": [32]
  }, undefined, {
    "value": ["watchPropHandler"]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["c-steps"];
  components.forEach(tagName => { switch (tagName) {
    case "c-steps":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, CSteps$1);
      }
      break;
  } });
}

const CSteps = CSteps$1;
const defineCustomElement = defineCustomElement$1;

export { CSteps, defineCustomElement };
