// @cscfi/csc-ui/dist/components/c-sub-navigation-item.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { d as defineCustomElement$3 } from './c-loader2.js';
import { d as defineCustomElement$2 } from './c-spinner2.js';

const cSubNavigationItemCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host(c-sub-navigation-item){--_c-sub-navigation-item-text-color:var(\n    --c-sub-navigation-item-text-color,\n    var(--c-text-body)\n  );--_c-sub-navigation-item-background-color:var(\n    --c-sub-navigation-item-background-color,\n    var(--c-transparent)\n  );--_c-sub-navigation-item-background-color-hover:var(\n    --c-sub-navigation-item-background-color-hover,\n    var(--c-primary-100)\n  );--_c-sub-navigation-item-text-color-active:var(\n    --c-sub-navigation-item-text-color-active,\n    var(--c-primary-600)\n  );--_c-sub-navigation-item-background-color-active:var(\n    --c-sub-navigation-item-background-color-active,\n    var(--c-white)\n  );--_c-sub-navigation-item-sub-item-background-color-active:var(\n    --c-sub-navigation-item-sub-item-background-color-active,\n    var(--c-primary-100)\n  );--_c-sub-navigation-item-sub-item-background-color-hover:var(\n    --c-sub-navigation-item-sub-item-background-color-hover,\n    var(--c-primary-100)\n  );--_c-sub-navigation-item-sub-item-background-color:var(\n    --c-sub-navigation-item-sub-item-background-color,\n    var(--c-transparent)\n  );--_c-sub-navigation-item-sub-item-text-color-active:var(\n    --c-sub-navigation-item-sub-item-text-color-active,\n    var(--c-primary-600)\n  );--_c-sub-navigation-item-sub-item-text-color:var(\n    --c-sub-navigation-item-sub-item-text-color,\n    var(--c-text-body)\n  )}:host(.c-sub-navigation-item--sub-level){--_c-sub-navigation-item-background-color:var(\n    --_c-sub-navigation-item-sub-item-background-color\n  );--_c-sub-navigation-item-text-color:var(\n    --_c-sub-navigation-item-sub-item-text-color\n  );--_c-sub-navigation-item-background-color-hover:var(\n    --_c-sub-navigation-item-sub-item-background-color-hover\n  );--_c-sub-navigation-item-text-color-active:var(\n    --_c-sub-navigation-item-sub-item-text-color-active\n  );--_c-sub-navigation-item-background-color-active:var(\n    --_c-sub-navigation-item-sub-item-background-color-active\n  )}.c-sub-navigation-item{align-items:center;background-color:var(--_c-sub-navigation-item-background-color);border-radius:4px;color:var(--_c-sub-navigation-item-text-color);cursor:pointer;display:flex;font-weight:400;line-height:46px;margin:0 8px;overflow:hidden;padding-left:34px;position:relative;transition:background-color 0.2s ease-in;user-select:none}.c-sub-navigation-item__wrapper{padding:2px 0}.c-sub-navigation-item__content{display:flex;align-items:center;overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.c-sub-navigation-item__content::before{background-color:var(--_c-sub-navigation-item-text-color-active);content:\"\";height:100%;left:0;position:absolute;top:0;transform:translateZ(0) translateX(-8px);transition:transform 0.2s ease-in-out;width:8px}.c-sub-navigation-item:hover{background-color:var(--_c-sub-navigation-item-background-color-hover);color:var(--_c-sub-navigation-item-text-color-active)}:host(.active) .c-sub-navigation-item{background-color:var(--_c-sub-navigation-item-background-color-active)}:host(.active) .c-sub-navigation-item__content::before{transform:translateZ(0) translateX(0)}::slotted(span){margin-right:8px;font-size:20px;line-height:1}:host(:host:focus),:host(:host:focus) .c-sub-navigation-item{outline:none}:host(:host:focus-visible) .c-sub-navigation-item{outline:2px var(--_c-sub-navigation-item-text-color-active) solid;outline-offset:2px}.visuallyhidden{border:0;clip:rect(0 0 0 0);height:1px;margin:-1px;overflow:hidden;padding:0;position:absolute;white-space:nowrap;width:1px}.c-sub-navigation-item__slot{overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.c-sub-navigation-item__slot slot{display:flex;gap:8px;align-items:center}";

const CSubNavigationItem$1 = proxyCustomElement(class CSubNavigationItem extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.active = undefined;
    this.focusable = false;
    this.href = undefined;
    this.target = null;
    this.loading = false;
  }
  _redirect(event) {
    if ((event instanceof KeyboardEvent && (event === null || event === void 0 ? void 0 : event.key) === 'Enter') ||
      event instanceof MouseEvent ||
      event instanceof PointerEvent) {
      event.stopPropagation();
      const sidenav = document.querySelector('c-side-navigation');
      sidenav.menuVisible = false;
      if (this.href) {
        if (this.target) {
          window.open(this.href, this.target);
        }
        else {
          window.location.href = this.href;
        }
      }
    }
  }
  render() {
    const a11y = {
      tabindex: this.focusable ? '0' : '-1',
      role: 'menuitem',
    };
    if (this.active) {
      a11y['aria-current'] = 'page';
    }
    return (h(Host, Object.assign({}, a11y, { class: { active: this.active }, onClick: (e) => this._redirect(e), onKeyDown: (e) => this._redirect(e) }), h("div", { class: "c-sub-navigation-item__wrapper" }, h("div", { class: "c-sub-navigation-item" }, h("div", { class: "c-sub-navigation-item__content" }, h("div", { class: "c-sub-navigation-item__slot" }, h("slot", null)), this.active && (h("span", { class: "visuallyhidden" }, ", Current page"))), h("c-loader", { size: 32, hide: !this.loading, style: { pointerEvents: 'none' } })))));
  }
  get element() { return this; }
  static get style() { return cSubNavigationItemCss; }
}, [1, "c-sub-navigation-item", {
    "active": [4],
    "focusable": [4],
    "href": [1],
    "target": [1],
    "loading": [4]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["c-sub-navigation-item", "c-loader", "c-spinner"];
  components.forEach(tagName => { switch (tagName) {
    case "c-sub-navigation-item":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, CSubNavigationItem$1);
      }
      break;
    case "c-loader":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "c-spinner":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const CSubNavigationItem = CSubNavigationItem$1;
const defineCustomElement = defineCustomElement$1;

export { CSubNavigationItem, defineCustomElement };
