// @cscfi/csc-ui/dist/components/c-text-field.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { r as mdiEye, s as mdiEyeOff, t as mdiCalendar } from './mdi.js';
import { d as defineCustomElement$3 } from './c-input2.js';
import { d as defineCustomElement$2 } from './c-message2.js';

const cTextFieldCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:root,:host{--_c-text-field-active-color:var(--c-text-field-active-color, var(--c-primary-600));--_c-text-field-inactive-color:var(--c-text-field-inactive-color, var(--c-tertiary-600));--_c-text-field-background-color:var(--c-text-field-background-color, var(--c-transparent));--_c-text-field-text-color:var(--c-text-field-text-color, var(--c-text-body));--_c-text-field-placeholder-color:var(--c-text-field-placeholder-color, var(--c-tertiary-500));--_c-text-field-shadow-active-color:var(--c-text-field-shadow-active-color, var(--_c-text-field-active-color));--_c-text-field-shadow-background-color:var(--c-text-field-shadow-background-color, var(--c-white));--_c-text-field-shadow-text-color:var(--c-text-field-shadow-text-color, var(--_c-text-field-text-color));--_c-text-field-shadow-placeholder-color:var(--c-text-field-shadow-placeholder-color, var(--_c-text-field-placeholder-color))}.c-input{--_c-input-active-color:var(--_c-text-field-active-color);--_c-input-inactive-color:var(--_c-text-field-inactive-color);--_c-input-background-color:var(--_c-text-field-background-color);--_c-input-text-color:var(--_c-text-field-text-color);--_c-input-placeholder-color:var(--_c-text-field-placeholder-color);--_c-input-shadow-active-color:var(--_c-text-field-shadow-active-color);--_c-input-shadow-background-color:var(--_c-text-field-shadow-background-color);--_c-input-shadow-text-color:var(--_c-text-field-shadow-text-color);--_c-input-shadow-placeholder-color:var(--_c-text-field-shadow-placeholder-color)}.c-input__password-toggle,.c-input__date-toggle{cursor:pointer;fill:currentColor;height:22px;width:22px}.c-input__password-toggle--disabled,.c-input__date-toggle--disabled{cursor:not-allowed;fill:var(--c-tertiary-500)}.c-input input[type=date]{opacity:0}.c-input input[type=date]::-webkit-calendar-picker-indicator,.c-input input[type=date]::-webkit-inner-spin-button{display:none;appearance:none}.c-input input[type=date]:focus,.c-input input[type=date].c-input__input--filled{opacity:1}";

const CTextField$1 = proxyCustomElement(class CTextField extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.changeValue = createEvent(this, "changeValue", 3);
    this.internals = this.attachInternals();
    this._originalType = '';
    this._handleChange = (event) => {
      if (this.type === 'number' && !event.target.value && event.key === '-') {
        return;
      }
      this.value = this.trimWhitespace
        ? event.target.value.trim()
        : event.target.value;
      this.changeValue.emit(this.value);
      this.internals.setFormValue(this.value);
    };
    this._handleBlur = (event) => {
      if (this.trimWhitespace) {
        const trimmedValue = event.target.value.trim();
        event.target.value = trimmedValue;
      }
    };
    this._togglePasswordVisibility = () => {
      if (this.disabled)
        return;
      this.type = this.type === 'password' ? 'text' : 'password';
    };
    this.autofocus = false;
    this.autocapitalize = '';
    this.autocorrect = '';
    this.autocomplete = '';
    this.disabled = false;
    this.hideDetails = false;
    this.hint = '';
    this.hostId = undefined;
    this.trimWhitespace = false;
    this.label = undefined;
    this.max = null;
    this.min = null;
    this.name = undefined;
    this.placeholder = undefined;
    this.readonly = false;
    this.required = false;
    this.rows = 1;
    this.shadow = false;
    this.step = null;
    this.type = 'text';
    this.valid = true;
    this.validate = false;
    this.validateOnBlur = false;
    this.validation = 'Required field';
    this.value = undefined;
    this.isFocused = false;
    this.labelWidth = 0;
    this.preSlotWidth = 0;
  }
  componentWillLoad() {
    CTextField._uniqueId += 1;
    this._originalType = this.type;
    this._inputId = `${(this.hostId ||
      this.label ||
      this.placeholder ||
      '').replace(/[^a-zA-Z0-9-_]/g, '')}_${CTextField._uniqueId}`;
    this.internals.setFormValue(this.value);
  }
  get isActive() {
    return !!this.value || this.isFocused;
  }
  get passwordIcon() {
    return this.type === 'password' ? mdiEye : mdiEyeOff;
  }
  _renderInputElement() {
    const props = {
      classes: {
        'c-input__input': true,
        'c-input__input--filled': !!this.value,
      },
      shared: {
        id: this._inputId,
        name: this.name,
        disabled: this.disabled,
        readonly: this.readonly,
        value: this.value,
        onInput: this._handleChange,
        onChange: this._handleChange,
        onBlur: this._handleBlur,
      },
      input: Object.assign(Object.assign(Object.assign({ type: this.type, min: this.min, max: this.max, step: this.step }, (!!this.autocomplete && { autocomplete: this.autocomplete })), (!!this.autocapitalize && { autocapitalize: this.autocapitalize })), (!!this.autocorrect && { autocorrect: this.autocorrect })),
      textArea: {
        rows: this.rows,
      },
    };
    const textInput = (h("input", Object.assign({ class: props.classes }, props.input, props.shared, { ref: (el) => (this._inputElement = el) })));
    const textArea = (h("textarea", Object.assign({ class: props.classes }, props.shared, props.textArea)));
    return this.rows > 1 ? textArea : textInput;
  }
  _isFirefox() {
    return !!navigator.userAgent.match(/firefox|fxios/i);
  }
  _renderDateToggle() {
    if (this._originalType !== 'date' || this._isFirefox())
      return;
    const classes = {
      'c-input__date-toggle': true,
      'c-input__date-toggle--disabled': this.disabled,
    };
    return (h("svg", { class: classes, viewBox: "0 0 24 24", onClick: this._toggleDatepicker.bind(this) }, h("path", { d: mdiCalendar })));
  }
  _toggleDatepicker() {
    var _a;
    (_a = this._inputElement) === null || _a === void 0 ? void 0 : _a.showPicker();
  }
  _renderPasswordToggle() {
    if (this._originalType !== 'password')
      return;
    const classes = {
      'c-input__password-toggle': true,
      'c-input__password-toggle--disabled': this.disabled,
    };
    return (h("svg", { class: classes, viewBox: "0 0 24 24", onClick: this._togglePasswordVisibility }, h("path", { d: this.passwordIcon })));
  }
  render() {
    return (h(Host, null, h("c-input", { autofocus: this.autofocus, disabled: this.disabled, "hide-details": this.hideDetails, hint: this.hint, id: this.hostId, "input-id": this._inputId, label: this.label, name: this.name, placeholder: this.placeholder, readonly: this.readonly, required: this.required, rows: this.rows, shadow: this.shadow, type: this.type, valid: this.valid, validate: this.validate, "validate-on-blur": this.validateOnBlur, validation: this.validation, value: this.value }, h("slot", { name: "pre", slot: "pre" }), this._renderInputElement(), this._renderPasswordToggle(), this._renderDateToggle(), h("slot", { name: "post" }))));
  }
  static get formAssociated() { return true; }
  get hiddenEl() { return this; }
  static get style() { return cTextFieldCss; }
}, [65, "c-text-field", {
    "autofocus": [4],
    "autocapitalize": [1],
    "autocorrect": [1],
    "autocomplete": [1],
    "disabled": [4],
    "hideDetails": [4, "hide-details"],
    "hint": [1],
    "hostId": [1, "id"],
    "trimWhitespace": [4, "trim-whitespace"],
    "label": [1],
    "max": [2],
    "min": [2],
    "name": [1],
    "placeholder": [1],
    "readonly": [4],
    "required": [4],
    "rows": [2],
    "shadow": [4],
    "step": [2],
    "type": [1],
    "valid": [4],
    "validate": [4],
    "validateOnBlur": [4, "validate-on-blur"],
    "validation": [1],
    "value": [1025],
    "isFocused": [32],
    "labelWidth": [32],
    "preSlotWidth": [32]
  }]);
CTextField$1._uniqueId = 0;
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["c-text-field", "c-input", "c-message"];
  components.forEach(tagName => { switch (tagName) {
    case "c-text-field":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, CTextField$1);
      }
      break;
    case "c-input":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "c-message":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const CTextField = CTextField$1;
const defineCustomElement = defineCustomElement$1;

export { CTextField, defineCustomElement };
