// @cscfi/csc-ui/dist/components/c-toasts.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { C as CToastType, a as CToastPosition } from './index2.js';
import { d as defineCustomElement$3 } from './c-icon2.js';
import { d as defineCustomElement$2 } from './c-toast2.js';

const cToastsCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host{display:grid;gap:12px;grid-template-columns:1fr;left:0;max-width:100%;min-width:30vw;padding:12px;pointer-events:none;position:fixed;right:0;width:640px;z-index:10000}:host(.absolute){position:absolute}:host(.bottom){bottom:0}:host(.top){top:0}:host(.right){justify-content:end;left:auto;right:0}:host(.left){justify-content:start;left:0;right:auto}:host(.center){justify-content:center;margin:0 auto}";

const CToasts$1 = proxyCustomElement(class CToasts extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this._getDefaultOptions = () => ({
      type: CToastType.Info,
      duration: 6000,
      persistent: false,
      indeterminate: false,
      position: CToastPosition.Fixed,
      progress: false,
      id: `c-toast-item-${CToasts._uniqueId}`,
    });
    this.absolute = false;
    this.horizontal = 'center';
    this.vertical = 'bottom';
    this.messages = [];
  }
  async addToast(message) {
    const customMessages = this.messages.filter((message) => message.custom);
    if (message.custom && customMessages.length > 0) {
      console.warn(`Custom toast messages are restricted to 1 visible message due to slot reflection limitations.`);
    }
    else {
      CToasts._uniqueId += 1;
      requestAnimationFrame(() => {
        const defaultOptions = this._getDefaultOptions();
        this.messages = [
          ...this.messages,
          Object.assign(Object.assign(Object.assign({}, defaultOptions), message), { duration: +(message === null || message === void 0 ? void 0 : message.duration) > 0
              ? +message.duration
              : defaultOptions.duration }),
        ];
      });
    }
  }
  async removeToast(id) {
    const toast = this.el.shadowRoot.querySelector(`#c-toast--${id}`);
    toast === null || toast === void 0 ? void 0 : toast.closeToast();
  }
  _onMessageClose(event) {
    this._removeMessage(event.detail.id);
  }
  _removeMessage(id) {
    const toast = this.el.shadowRoot.querySelector(`#c-toast--${id}`);
    toast === null || toast === void 0 ? void 0 : toast.remove();
    const messageCount = this.el.shadowRoot.querySelectorAll('c-toast').length;
    if (messageCount === 0) {
      this.messages = [].slice();
    }
  }
  _renderMessage(message) {
    return (h("c-toast", { message: message, onClose: (e) => this._onMessageClose(e) }, message.custom && h("slot", null)));
  }
  componentWillLoad() {
    CToasts._uniqueId += 1;
  }
  render() {
    return (h(Host, { class: {
        absolute: this.absolute,
        [this.vertical]: true,
        [this.horizontal]: true,
      } }, this.messages.map((message) => this._renderMessage(message))));
  }
  get el() { return this; }
  static get style() { return cToastsCss; }
}, [1, "c-toasts", {
    "absolute": [4],
    "horizontal": [1],
    "vertical": [1],
    "messages": [32],
    "addToast": [64],
    "removeToast": [64]
  }]);
CToasts$1._uniqueId = 0;
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["c-toasts", "c-icon", "c-toast"];
  components.forEach(tagName => { switch (tagName) {
    case "c-toasts":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, CToasts$1);
      }
      break;
    case "c-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "c-toast":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const CToasts = CToasts$1;
const defineCustomElement = defineCustomElement$1;

export { CToasts, defineCustomElement };
