// @cscfi/csc-ui/dist/collection/styles/tailwind/theme.js
export const theme = {
  colors: {
    accent: {
      100: "#e5f9f7ff",
      200: "#ccf4f0ff",
      300: "#99e9e0ff",
      400: "#66ddd1ff",
      500: "#33d2c1ff",
      600: "#00c7b2ff",
      700: "#009f8eff",
      800: "#00776bff",
      900: "#005047ff",
      DEFAULT: "#00c7b2ff",
    },
    black: "#000000ff",
    error: {
      100: "#fce7e7ff",
      200: "#f8ceceff",
      300: "#f19e9eff",
      400: "#eb6d6dff",
      500: "#e43d3dff",
      600: "#dd0c0cff",
      700: "#ae0909ff",
      800: "#7e0707ff",
      900: "#4d0404ff",
      DEFAULT: "#dd0c0cff",
    },
    info: {
      100: "#e5f2f8ff",
      200: "#cce6f1ff",
      300: "#99cde4ff",
      400: "#66b4d6ff",
      500: "#339bc9ff",
      600: "#0082bbff",
      700: "#006896ff",
      800: "#004e70ff",
      900: "#00344bff",
      DEFAULT: "#0082bbff",
    },
    link: {
      100: "#e5eaefff",
      200: "#ccd5dfff",
      300: "#99acbfff",
      400: "#66829fff",
      500: "#33597fff",
      600: "#002f5fff",
      700: "#00264cff",
      800: "#001c39ff",
      900: "#001326ff",
      DEFAULT: "#002f5fff",
    },
    primary: {
      100: "#e6f0f2ff",
      200: "#cce1e4ff",
      300: "#99c2c9ff",
      400: "#66a4aeff",
      500: "#338593ff",
      600: "#006778ff",
      700: "#005260ff",
      800: "#003e48ff",
      900: "#002930ff",
      DEFAULT: "#006778ff",
    },
    secondary: {
      100: "#f3e5eeff",
      200: "#e6ccdcff",
      300: "#cd99b9ff",
      400: "#b56697ff",
      500: "#9c3374ff",
      600: "#830051ff",
      700: "#690041ff",
      800: "#4f0031ff",
      900: "#340020ff",
      DEFAULT: "#830051ff",
    },
    success: {
      100: "#edf7e7ff",
      200: "#dceeceff",
      300: "#b9dc9cff",
      400: "#97cb6bff",
      500: "#74b939ff",
      600: "#51a808ff",
      700: "#418606ff",
      800: "#316505ff",
      900: "#204303ff",
      DEFAULT: "#51a808ff",
    },
    tertiary: {
      100: "#eff0f1ff",
      200: "#dfe1e3ff",
      300: "#bfc3c6ff",
      400: "#9ea6aaff",
      500: "#7e888dff",
      600: "#5e6a71ff",
      700: "#4b555aff",
      800: "#384044ff",
      900: "#262a2dff",
      DEFAULT: "#5e6a71ff",
    },
    dark: "#2e3438ff",
    light: "#ffffffff",
    transparent: "transparent",
    warning: {
      100: "#ffeee5ff",
      200: "#ffdeccff",
      300: "#ffbc99ff",
      400: "#ff9b66ff",
      500: "#ff7933ff",
      600: "#ff5800ff",
      700: "#cc4600ff",
      800: "#993500ff",
      900: "#662300ff",
      DEFAULT: "#ff5800ff",
    },
    white: "#ffffffff",
    "font-family": "'museo-sans', sans-serif",
    "text-body": "var(--c-primary-900)",
    "text-system": "var(--c-tertiary-600)",
  },
};
