// @cscfi/csc-ui/dist/collection/components/c-spinner/c-spinner.js
import { Host, h } from "@stencil/core";
export class CSpinner {
  constructor() {
    this.color = 'var(--c-primary-600)';
    this.size = 24;
    this.width = 2;
  }
  render() {
    const iconParams = {
      cx: this.size / 2,
      cy: this.size / 2,
      r: this.size / 2 - 2,
    };
    const style = {
      '--c-size': `${this.size}px`,
      '--c-width': `${this.width}px`,
      '--c-color': this.color,
    };
    return (h(Host, { class: "c-spinner", style: style }, h("svg", null, h("circle", Object.assign({}, iconParams)))));
  }
  static get is() { return "c-spinner"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["c-spinner.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["c-spinner.css"]
    };
  }
  static get properties() {
    return {
      "color": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Color of the spinner"
        },
        "attribute": "color",
        "reflect": false,
        "defaultValue": "'var(--c-primary-600)'"
      },
      "size": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Size of the spinner"
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "24"
      },
      "width": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Width of the spinner"
        },
        "attribute": "width",
        "reflect": false,
        "defaultValue": "2"
      }
    };
  }
}
