// @cscfi/csc-ui/dist/esm/c-swiper-tab.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-47956d53.js';

const cSwiperTabCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host(c-swiper-tab){--_c-swiper-tab-background-color-active:var(--c-swiper-tab-background-color-active, var(--c-primary-600));--_c-swiper-tab-background-color-disabled:var(--c-swiper-tab-background-color-disabled, var(--c-tertiary-100));--_c-swiper-tab-background-color:var(--c-swiper-tab-background-color, var(--c-primary-200));--_c-swiper-tab-hover-color:var(--c-swiper-tab-hover-color, var(--c-primary-100));--_c-swiper-tab-text-color-active:var(--c-swiper-tab-text-color-active, var(--c-white));--_c-swiper-tab-text-color-disabled:var(--c-swiper-tab-text-color-disabled, var(--c-tertiary-600));--_c-swiper-tab-text-color:var(--c-swiper-tab-text-color, var(--c-primary-600));border-radius:8px;width:100%;height:100%;outline:none;text-decoration:none;user-select:none;display:grid;align-content:stretch;padding:4px;background-color:transparent;border:none}:host(:focus){outline:2px var(--_c-swiper-tab-background-color-active) solid;outline-offset:-2px}@supports selector(:focus-visible){:host(:focus){outline:none}}:host(:focus-visible){outline:2px var(--_c-swiper-tab-background-color-active) solid;outline-offset:-2px}.c-swiper-tab{height:100%}.c-swiper-tab:hover:not(.c-swiper-tab--active):not(.c-swiper-tab--disabled) .c-swiper-tab__content{background-color:var(--_c-swiper-tab-hover-color)}.c-swiper-tab__content{border-radius:6px;background-color:var(--_c-swiper-tab-background-color);color:var(--_c-swiper-tab-text-color);cursor:pointer;transition:background-color 0.2s ease-in-out;display:grid;gap:4px;grid-template-columns:1fr;line-height:1.5;padding:16px;align-content:start;position:relative;overflow:hidden;height:100%}.c-swiper-tab__header{line-height:38px;font-size:24px;display:grid;grid-template-columns:1fr auto;gap:16px}.c-swiper-tab__description{align-items:start;opacity:0.67}.c-swiper-tab--disabled .c-swiper-tab__content{background:var(--_c-swiper-tab-background-color-disabled);color:var(--_c-swiper-tab-text-color-disabled);cursor:default !important;pointer-events:none}.c-swiper-tab--active .c-swiper-tab__content{background-color:var(--_c-swiper-tab-background-color-active);color:var(--_c-swiper-tab-text-color-active);pointer-events:none}.c-swiper-tab--active .c-swiper-tab__content:hover{background-color:var(--_c-swiper-tab-background-color-active)}.c-swiper-tab svg{fill:currentColor;height:38px;width:38px}";

const CSwiperTab = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.changeValue = createEvent(this, "changeValue", 7);
    this.disabled = false;
    this.active = false;
    this.label = undefined;
    this.hostId = undefined;
    this.setsize = undefined;
    this.position = undefined;
    this.value = undefined;
  }
  onTabClick(event) {
    if (this.active || this.disabled)
      return;
    this._rippleElement.createRipple(event, this._container);
    this.changeValue.emit(this.value);
  }
  render() {
    const classes = {
      'c-swiper-tab': true,
      'c-swiper-tab--active': !this.disabled && this.active,
      'c-swiper-tab--disabled': this.disabled,
    };
    const a11y = {
      'aria-selected': this.active ? 'true' : 'false',
      'aria-setsize': this.setsize,
      'aria-posinset': this.position,
      tabindex: this.active ? 0 : -1,
    };
    return (h(Host, Object.assign({}, a11y, { role: "tab" }), h("div", { id: this.hostId, class: classes }, h("div", { class: "c-swiper-tab__content", ref: (el) => (this._container = el) }, h("div", { class: "c-swiper-tab__header" }, this.label, h("slot", { name: "icon" })), h("div", { class: "c-swiper-tab__description" }, h("slot", null)), h("c-ripple", { ref: (el) => (this._rippleElement = el) })))));
  }
  get el() { return getElement(this); }
};
CSwiperTab.style = cSwiperTabCss;

export { CSwiperTab as c_swiper_tab };
