// @cscfi/csc-ui/dist/esm/c-toast.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-47956d53.js';
import { a as mdiClose, e as mdiAlert, f as mdiCloseCircle, n as mdiCheckCircle, d as mdiInformation } from './mdi-18a39071.js';

const cToastCss = "html{box-sizing:border-box}body{margin:0}*,*:before,*:after{box-sizing:inherit;-webkit-tap-highlight-color:transparent}:host{font-family:var(--c-font-family)}::-ms-reveal{display:none}:host(c-toast){--_c-toast-background-color:var(--c-toast-background-color, var(--c-white));--_c-toast-text-color:var(--c-toast-text-color, var(--c-text-system));--_c-toast-animation-state:running;align-items:center;background-color:var(--_c-toast-background-color);border-radius:6px;border:2px solid var(--_c-toast-color);border-left-width:12px;box-shadow:rgba(0, 0, 0, 0.07) 0 1px 1px, rgba(0, 0, 0, 0.07) 0 2px 2px, rgba(0, 0, 0, 0.07) 0 4px 4px, rgba(0, 0, 0, 0.07) 0 8px 8px, rgba(0, 0, 0, 0.07) 0 16px 16px;box-sizing:border-box;color:var(--_c-toast-text-color);display:grid;min-height:52px;opacity:0;padding:8px 12px;pointer-events:all;text-align:left;transform:translateY(20px);transition:opacity 0.5s ease, transform 0.5s ease;width:100%;will-change:opacity, transform;z-index:10001;position:relative}:host(.show){opacity:1;transform:translateY(0)}:host(.info){--_c-toast-color:var(--c-info-600)}:host(.error){--_c-toast-color:var(--c-error-600)}:host(.warning){--_c-toast-color:var(--c-warning-600)}:host(.success){--_c-toast-color:var(--c-success-600)}.c-toast__content p{margin:0;font-weight:600}.c-toast__item{align-items:center;display:grid;gap:12px;grid-auto-columns:auto;grid-auto-flow:column;grid-template-columns:24px 1fr;font-weight:300}.c-toast__item>svg{fill:var(--_c-toast-color)}.c-toast__item svg{height:24px;width:24px}.c-toast__progress{background-color:var(--c-tertiary-200);border-radius:8px;height:6px;margin-top:8px;overflow:hidden;transform:translateZ(0)}:host(:hover){--_c-toast-animation-state:paused}.c-toast__progress__bar{animation-duration:var(--_c-toast-duration);animation-fill-mode:forwards;animation-iteration-count:1;animation-name:timer;animation-play-state:var(--_c-toast-animation-state);animation-timing-function:linear;background-color:var(--_c-toast-color);border-radius:8px;height:6px;width:100%}.c-toast__progress__bar.indeterminate{animation:indeterminate 2s infinite linear;left:-100%;width:50%;position:relative;transform-origin:0% 50%}.visuallyhidden{border:0;clip:rect(1px, 1px, 1px, 1px);height:1px;overflow:hidden;padding:0;position:absolute;width:1px}@keyframes timer{100%{transform:translateX(-100%)}}@keyframes indeterminate{0%{transform:translateX(100%) scaleX(1)}30%{transform:translateX(170%) scaleX(1.75)}70%{transform:translateX(500%) scaleX(0.1)}100%{transform:translateX(500%) scaleX(0.1)}}";

const CToast = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.close = createEvent(this, "close", 7);
    this._closed = false;
    this._icons = {
      close: mdiClose,
      warning: mdiAlert,
      error: mdiCloseCircle,
      success: mdiCheckCircle,
      info: mdiInformation,
    };
    this.message = undefined;
  }
  async closeToast() {
    if (this._closed)
      return;
    this._closed = true;
    this.el.classList.remove('show');
    this.el.ontransitionend = () => this.close.emit(this.message);
  }
  componentDidLoad() {
    window.requestAnimationFrame(() => {
      this.el.classList.add('show');
    });
    if (this.message.persistent || this.message.indeterminate)
      return;
    this._startTime = Date.now();
    this._remainingTime = this.message.duration;
    this._timeout = window.setTimeout(() => this.closeToast(), +this.message.duration);
  }
  disconnectedCallback() {
    window.clearTimeout(this._timeout);
  }
  _onMouseEnter() {
    if (this.message.persistent || this.message.indeterminate)
      return;
    this._remainingTime = this._remainingTime - (Date.now() - this._startTime);
    window.clearTimeout(this._timeout);
  }
  _onMouseLeave() {
    if (this.message.persistent || this.message.indeterminate)
      return;
    this._startTime = Date.now();
    this._timeout = window.setTimeout(() => this.closeToast(), this._remainingTime);
  }
  _renderCloseButton() {
    const Tag = !!this.message.closeText ? 'c-button' : 'c-icon-button';
    return (h(Tag, { "aria-label": "close", size: "small", text: true, onClick: () => this.closeToast() }, h("c-icon", Object.assign({}, (!!this.message.closeText && { slot: 'icon' }), { color: "var(--_c-toast-color)", path: this._icons.close })), this.message.closeText));
  }
  render() {
    const showProgressBar = !this.message.persistent && this.message.progress;
    return (h(Host, { id: `c-toast--${this.message.id}`, class: { [this.message.type]: true }, role: "alert", "aria-atomic": "true", "aria-live": "assertive", onMouseEnter: () => this._onMouseEnter(), onMouseLeave: () => this._onMouseLeave() }, h("span", { class: "visuallyhidden" }, this.message.type, " notification"), this.message.custom ? (h("div", { class: "c-toast__custom-item" }, h("div", { class: "c-toast__content" }, h("slot", null)))) : (h("div", { class: "c-toast__item" }, h("svg", { viewBox: "0 0 24 24" }, h("path", { d: this._icons[this.message.type] })), h("div", { class: "c-toast__content" }, !!this.message.title && h("p", null, this.message.title), this.message.message), !this.message.indeterminate && this._renderCloseButton())), showProgressBar && (h("div", { class: "c-toast__progress", style: { '--_c-toast-duration': `${this.message.duration}ms` } }, h("div", { class: {
        'c-toast__progress__bar': true,
        indeterminate: this.message.indeterminate,
      } })))));
  }
  get el() { return getElement(this); }
};
CToast.style = cToastCss;

export { CToast as c_toast };
